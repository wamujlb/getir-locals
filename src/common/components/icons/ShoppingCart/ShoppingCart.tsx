import React from 'react';
import { SvgIcon, SvgIconProps } from '@mui/material';

const ShoppingCartIcon = (props: SvgIconProps) => {
  return (
    <SvgIcon {...props}>
      <path d="M3.9,7.8h16.2V22H3.9V7.8z M9.1,1.9h5.8L16,3.1v4.7l-1.2,0V3.1H9.2v4.7L8,7.8V3.1L9.1,1.9z" />
    </SvgIcon>
  );
};

ShoppingCartIcon.displayName = 'ShoppingCartIcon';

export default ShoppingCartIcon;
