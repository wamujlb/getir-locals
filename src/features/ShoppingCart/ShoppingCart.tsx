import React, { useCallback } from 'react';
import { Form, Formik } from 'formik';

import ShoppingCartCard from 'common/components/ShoppingCart/ShoppingCartCard/ShoppingCartCard';
import ShoppingCartItem from 'common/components/ShoppingCart/ShoppingCartItem/ShoppingCartItem';
import { useShoppingCartForm } from './hooks/useShoppingCardForm';

const ShoppingCart = () => {
  const { initialValues, onSubmit } = useShoppingCartForm();

  return (
    <ShoppingCartCard totalPrice={12321} variant="outlined">
      <Formik initialValues={initialValues} onSubmit={onSubmit}>
        {() => (
          <Form>
            <ShoppingCartItem
              id="id_1"
              name="items.id_1"
              title="Example Product"
              price={14.99}
              imageUri=""
            />

            <ShoppingCartItem
              id="id_2"
              name="items.id_2"
              title="Example Product"
              price={14.99}
              imageUri=""
            />

            <ShoppingCartItem
              id="id_3"
              name="items.id_3"
              title="Example Product"
              price={14.99}
              imageUri=""
            />
          </Form>
        )}
      </Formik>
    </ShoppingCartCard>
  );
};

ShoppingCart.displayName = 'ShoppingCart';

export default ShoppingCart;
