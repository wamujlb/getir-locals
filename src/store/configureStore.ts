import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';

import rootReducer from './reducers';
import { watcherSaga } from './sagas/rootSaga';

export type RootState = ReturnType<typeof rootReducer>;
export type BaseThunkApiConfig = { state: RootState };

const sagaMiddleware = createSagaMiddleware();

export const createStore = (preloadedState: PreloadedState<RootState> = {}) =>
  configureStore({
    preloadedState,
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) => [
      ...getDefaultMiddleware({ thunk: false }),
      sagaMiddleware,
    ],
    devTools: process.env.NODE_ENV === 'development',
  });

const store = createStore();

sagaMiddleware.run(watcherSaga);

export type AppDispatch = typeof store.dispatch;

export type AppStore = typeof store;

export default store;
