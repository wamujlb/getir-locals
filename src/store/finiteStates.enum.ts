export enum FiniteStates {
  Idle = 'Idle',
  Loading = 'Loading',
  Fetching = 'Fetching',
  Success = 'Success',
  Failure = 'Failure',
}
