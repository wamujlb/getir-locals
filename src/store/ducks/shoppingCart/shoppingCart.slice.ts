import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from 'store/configureStore';
import { ProductItemData } from 'features/ProductResults/ProductResults.interface';

export type ShoppingCartItemData = {
  count: number;
} & ProductItemData;

export type ShoppingCartItems = {
  [key: string]: ShoppingCartItemData;
};

export type ShoppingCartState = {
  data: {
    items: ShoppingCartItems;
  };
};

export const initialState: ShoppingCartState = {
  data: {
    items: {},
  },
};

export const shoppingCartSlice = createSlice({
  name: 'shoppingCart',
  initialState,
  reducers: {
    addItem: (state, { payload }: PayloadAction<ProductItemData>) => {
      state.data.items = {
        ...state.data.items,
        [payload.id]: {
          ...payload,
          count: 1,
        },
      };
    },
    updateItemCount: (
      state,
      { payload }: PayloadAction<{ id: string; newCount: number }>
    ) => {
      const item: ShoppingCartItemData | undefined =
        state.data.items[payload.id];

      if (item) {
        // Remove item when newCount is 0
        if (payload.newCount === 0) {
          delete state.data.items[payload.id];
          return;
        }

        state.data.items = {
          ...state.data.items,
          [payload.id]: {
            ...item,
            count: payload.newCount,
          },
        };
      }
    },
    removeItem: (state, { payload }: PayloadAction<string>) => {
      if (payload in state.data.items) {
        delete state.data.items[payload];
      }
    },
    clearShoppingCart: (state) => {
      state.data.items = {};
    },
  },
});

const shoppingCartDataSelector = (
  state: RootState
): ShoppingCartState['data'] => state.shoppingCart.data;

export const selectShoppingCartItems = createSelector(
  shoppingCartDataSelector,
  (data): ShoppingCartItems => data.items
);

export const selectShoppingCartItemsList = createSelector(
  shoppingCartDataSelector,
  (data): ShoppingCartItemData[] => Object.values(data.items)
);

export const selectShoppingCartTotalPrice = createSelector(
  selectShoppingCartItemsList,
  (items: ShoppingCartItemData[]): number => {
    return items.reduce(
      (acc, { price, count }: ShoppingCartItemData) => acc + price * count,
      0
    );
  }
);

export const { addItem, updateItemCount, removeItem, clearShoppingCart } =
  shoppingCartSlice.actions;

export default shoppingCartSlice.reducer;
