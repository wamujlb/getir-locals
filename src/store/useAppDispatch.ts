import { useDispatch } from 'react-redux';

import { AppDispatch } from './configureStore';

export const useAppDispatch = () => useDispatch<AppDispatch>();
