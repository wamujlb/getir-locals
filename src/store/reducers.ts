import { combineReducers } from 'redux';

import shoppingCart from 'store/ducks/shoppingCart/shoppingCart.slice';

export default combineReducers({
  shoppingCart,
});
