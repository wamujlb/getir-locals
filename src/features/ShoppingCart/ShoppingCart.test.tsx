import React from 'react';
import { render, cleanup } from '@testing-library/react';

import ShoppingCart from './ShoppingCart';
import {
  WrappedWithRedux,
  WrappedWithReduxProps,
  WrappedWithTheme,
} from 'test-helpers';
import { createStore } from 'store/configureStore';
import {
  productItemDataMock,
  productItemDataMock_2,
} from 'features/ProductResults/__mocks__/productItem.mock';
import userEvent from '@testing-library/user-event';
import { addItem } from 'store/ducks/shoppingCart/shoppingCart.slice';

const TestComponent = ({ store }: WrappedWithReduxProps) => {
  return (
    <WrappedWithRedux store={store}>
      <WrappedWithTheme>
        <ShoppingCart />
      </WrappedWithTheme>
    </WrappedWithRedux>
  );
};

describe('#ShoppingCart', () => {
  const store = createStore();

  afterEach(() => {
    cleanup();
  });

  it('should match the snapshot when shopping cart is empty', () => {
    const view = render(<TestComponent store={store} />);

    expect(view.container).toMatchSnapshot();
  });

  it('should show 1 item in the cart', () => {
    store.dispatch(addItem(productItemDataMock));

    const view = render(<TestComponent store={store} />);

    expect(view.container).toMatchSnapshot();
  });

  it('should remove the item from the shopping cart when the count becomes 0', () => {
    const view = render(<TestComponent store={store} />);

    expect(view.queryByRole('spinbutton')).toHaveValue(1);

    userEvent.click(view.getByRole('button', { name: 'Decrease count' }));

    expect(view.queryByRole('spinbutton')).not.toBeInTheDocument();
  });

  it('should change the count of the item in the cart', () => {
    store.dispatch(addItem(productItemDataMock));

    const view = render(<TestComponent store={store} />);

    expect(view.queryByRole('spinbutton')).toHaveValue(1);

    userEvent.click(view.getByRole('button', { name: 'Increase count' }));

    userEvent.click(view.getByRole('button', { name: 'Increase count' }));

    expect(view.getByRole('spinbutton')).toHaveValue(3);
  });
});
