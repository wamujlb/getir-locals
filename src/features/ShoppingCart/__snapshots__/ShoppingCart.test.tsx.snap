// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#ShoppingCart should match the snapshot when shopping cart is empty 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded makeStyles-root-1 MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiBox-root css-19midj6"
    >
      <ul
        class="MuiList-root css-1mk9mw3-MuiList-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1uwgr7b-MuiTypography-root"
        >
          No items
        </p>
      </ul>
      <div
        class="MuiBox-root css-x85yq3"
      >
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1mk3q3f-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          ₺ 0.00
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`#ShoppingCart should show 1 item in the cart 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded makeStyles-root-2 MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiBox-root css-19midj6"
    >
      <ul
        class="MuiList-root css-1mk9mw3-MuiList-root"
      >
        <li
          class="makeStyles-root-3 MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-divider css-h8lxwn-MuiListItem-root"
        >
          <div
            class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-primary css-et1ao3-MuiTypography-root"
              title="Gorgeous Office Mug"
            >
              Gorgeous Office Mug
            </span>
            <p
              class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-xomozu-MuiTypography-root"
            >
              ₺ 10.50
            </p>
          </div>
          <div
            class="MuiFormControl-root makeStyles-root-5 css-1nrlq1o-MuiFormControl-root"
          >
            <button
              aria-label="Decrease count"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="RemoveRoundedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 13H6c-.55 0-1-.45-1-1s.45-1 1-1h12c.55 0 1 .45 1 1s-.45 1-1 1z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <input
              class="makeStyles-input-6"
              name="a"
              readonly=""
              type="number"
              value="1"
            />
            <button
              aria-label="Increase count"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-2rshpx-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                data-testid="AddRoundedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
      </ul>
      <div
        class="MuiBox-root css-x85yq3"
      >
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1mk3q3f-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          ₺ 10.50
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
